[v9] Add in app access connection monitoring.

mdwn opened the pull request:
Backport https://github.com/gravitational/teleport/pull/17436 to branch/v9

Note: This ended up needing a good amount of rework because Teleport didn't support TCP apps here. While at face value you might think that we could avoid having a connection monitor here, websockets still benefit from the connection monitor since they're long lived.

The integration tests have been re-oriented to target websockets instead of TCP applications. The certification duration used `apidefaults.CertDuration`, so the only way to alter this without significantly altering a lot of the underlying plumbing was to set a max TTL for the role associated with the user.

Rather than track errors in the application server like the v10 and v11 implementations, this actually tracks whether a monitor was set up for a given connection and sets it up in the HTTP server itself. This is necessary because the function signature of `HandleConnection` doesn't include an error, so there's no real way to pass errors up the stack if authentication fails during `HandleConnection`.

Basically, please take a good look at this one since it's changed a good bit from the other backports.

mdwn committed:
Add in app access connection monitoring.

Application access connection monitoring has been introduced so that, when a
lock is created, application access connections will be interrupted until the
lock has been cleared. This includes web sockets and TCP applications.

r0mant reviewed (changes requested):
Do not merge until new tests flakiness is addressed.

mdwn committed:
Make TestTCP* tests in appaccess more deterministic. (#18233)

The appaccess TestTCP* tests are highly reliant on time. This has been
reduced (but not eliminated) by using a fakeClock and a channel for
signaling monitor triggered connection closures.
